Dynamic batch size demo (#745)
openvinotoolkit/model_server: docs/dynamic_batch_size.md

# Dynamic batch size with OpenVINO&trade; Model Server Demultiplexing

## Intoduction

> This documentation presents how to use OpenVINO&trade; Model Server to execute inference requests with dymanic batch size using [Demultiplexing in Directed Acyclic Graph Scheduler](https://github.com/openvinotoolkit/model_server/blob/develop/docs/demultiplexing.md).

Reviewer: This document guides how to configure DAG Scheduler pipeline to be able to send predict request with arbitrary batch size without model reloading.

With OpenVINO&trade; Model Server Demultiplexing infer request sent from client application can have various batch sizes and changing batch size does not require model reload.

*Note:* When using `demultiply_count` parameters, only one demultiplexer can exist in pipeline.

> - The script [grpc_serving_client.py](../example_client/grpc_serving_client.py) runs desired model inference requests for the images in .npy file.

Reviewer: Example client in python grpc_serving_client.py can be used to request the pipeline. Use `--dag-batch-size-auto` flag to add additional dimension to the input shape which is required for demultiplexing feature.

I would also move client part to the bottom

> - The example relies on the model [resnet](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/intel/resnet50-binary-0001/README.md).

Reviewer: Maybe "uses" would be better word?

> - While using resnet model with grpc_serving_client.py the script processes the output from the server and displays the inference results using previously prepared file with labels. Inside this file each image in input is assigned to a certain number, which indicates the correct recognision answer. 

Reviewer 2: ```suggestion
- While using resnet model with grpc_serving_client.py the script processes the output from the server and displays the inference results using previously prepared file with labels. Inside this file each image has assigned a certain number, which indicates the correct recognition answer. 
```

## Steps
Clone OpenVINO&trade; Model Server github repository and enter `model_server` directory.
#### Download the pretrained model

> Download both components of the model(xml and bin file) using curl and store it in `models` directory

Reviewer 2: ```suggestion
Download model files and store it in `models` directory
```

```Bash
mkdir -p models/resnet/1
curl https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/resnet50-binary-0001/FP32-INT1/resnet50-binary-0001.bin https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/resnet50-binary-0001/FP32-INT1/resnet50-binary-0001.xml -o models/resnet/1/resnet50-binary-0001.bin -o models/resnet/1/resnet50-binary-0001.xml
```

> #### Pull and tag the docker image with OVMS component

Reviewer: There is no tagging process, I think you meant "Pull the latest OVMS image from dockerhub"

Pull the latest version of OpenVINO&trade; Model Server from Dockerhub :
```Bash
docker pull openvino/model_server:latest
```

#### OVMS configuration file
Create new file named `config.json` :
```json
{
   "model_config_list": [
       {
           "config": {
               "name": "resnet",
               "base_path": "/models/resnet"
           }
       }
   ],
   "pipeline_config_list": [
       {
           "name": "resnet50DAG",
           "inputs": [
               "0"
           ],

>            "demultiply_count" : 0,

Reviewer: We already have such config json presented here:
https://github.com/openvinotoolkit/model_server/blob/main/docs/demultiplexing.md#dynamic-batch-handling-with-demultiplexing

Why do we need to have separate doc here? @dtrawins  @mkuczyns11

imagenet top results in a single batch:
         0 magnetic compass 635 ; Correct match.
Iteration 7; Processing time: 45.71 ms; speed 21.88 fps
imagenet top results in a single batch:
         0 peacock 84 ; Correct match.
Iteration 8; Processing time: 45.31 ms; speed 22.07 fps
imagenet top results in a single batch:
         0 pelican 144 ; Correct match.
Iteration 9; Processing time: 44.89 ms; speed 22.28 fps
imagenet top results in a single batch:
         0 snail 113 ; Correct match.
Iteration 10; Processing time: 46.63 ms; speed 21.45 fps
imagenet top results in a single batch:
         0 zebra 340 ; Correct match.

processing time for all iterations
average time: 45.80 ms; average speed: 21.83 fps
median time: 45.00 ms; median speed: 22.22 fps
max time: 50.00 ms; min speed: 20.00 fps
min time: 43.00 ms; max speed: 23.26 fps
time percentile 90: 48.20 ms; speed percentile 90: 20.75 fps
time percentile 50: 45.00 ms; speed percentile 50: 22.22 fps
time standard deviation: 1.94
time variance: 3.76
Classification accuracy: 100.00
```
Each iteration presents the results of each infer request and details for each image in batch.

> #### Run the client with different batch size :

Reviewer: Space before :

More information about this feature can be found in [dynamic batch size in demultiplexing](demultiplexing.md#dynamic-batch-handling-with-demultiplexing)

*Note:* When using `demultiply_count` parameters, only one demultiplexer can exist in pipeline.

- Example client in python grpc_serving_client.py can be used to request the pipeline. Use `--dag-batch-size-auto` flag to add additional dimension to the input shape which is required for demultiplexing feature.

- The example uses model [resnet](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/intel/resnet50-binary-0001/README.md).

- While using resnet model with grpc_serving_client.py the script processes the output from the server and displays the inference results using previously prepared file with labels. Inside this file each image has assigned number, which indicates the correct recognition answer.  

## Steps
Clone OpenVINO&trade; Model Server github repository and enter `model_server` directory.
#### Download the pretrained model
Download model files and store it in `models` directory
```Bash
mkdir -p models/resnet/1

> curl https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/resnet50-binary-0001/FP32-INT1/resnet50-binary-0001.bin https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/1/resnet50-binary-0001/FP32-INT1/resnet50-binary-0001.xml -o models/resnet/1/resnet50-binary-0001.bin -o models/resnet/1/resnet50-binary-0001.xml

Reviewer 3: new location for the models is in https://storage.openvinotoolkit.org/repositories/open_model_zoo/2021.4/models_bin/2/resnet50-binary-0001/FP32-INT1/

           }
       }
   ],
   "pipeline_config_list": [
       {
           "name": "resnet50DAG",
           "inputs": [
               "0"
           ],
           "demultiply_count" : 0,
           "nodes": [
               {
                   "name": "resnetNode",
                   "model_name": "resnet",
                   "type": "DL model",
                   "inputs": [
                       {
                           "0": {
                               "node_name": "request",
                               "data_item": "0"
                           }
                       }
                   ],
                   "outputs": [
                       {
                           "data_item": "1463",
                           "alias": "1463"
                       }
                   ]
               }
           ],
           "outputs": [
               {"1463": {
                       "node_name": "resnetNode",
                       "data_item": "1463"}}
           ]
       }
   ]
}
```

#### Start ovms docker container with downloaded model
Start ovms container with image pulled in previous step and mount `models` directory :
```Bash

> docker run --rm -d -v $(pwd)/models:/models -v $(pwd)/config.json:/config.json -p 9000:9000 openvino/model_server:latest --config_path config.json --port 9000

Reviewer 3: add in the command plugin_config setting. Number of streams is particularly important here

Author: Correct me if I am wrong, but adding it in the command is not possible due to running ovms with configuration file. However, I am going to add this setting to config.

```

> #### Run the client :

Reviewer: Remove `:`

```Bash
cd example_client
virtualenv .venv
. .venv/bin/activate
pip install -r client_requirements.txt

> python grpc_serving_client.py --images_numpy_path imgs.npy --labels_numpy_path lbs.npy --input_name 0 --output_name 1463 --model_name resnet50DAG --dag-batch-size-auto --transpose_input False --batchsize 1 ; python grpc_serving_client.py --images_numpy_path imgs.npy --labels_numpy_path lbs.npy --input_name 0 --output_name 1463 --model_name resnet50DAG --dag-batch-size-auto --transpose_input False --batchsize 8;

Reviewer 3: It might be easier to read when the output is directly after each command.

Author: This is done in such a way because I intended to show that there is no need to reload the config. Should I split the commands as You are suggesting?

Reviewer 3: It might be better to start the clients in background to ensure they run simultaneously with & sign.
The output should be clearly explained and separated to show which part is for which BS.

```
#### Output of the script
```Bash
...
Start processing:
        Model name: resnet50DAG
        Iterations: 10
        Images numpy path: imgs.npy

>         Images in shape: (10, 3, 224, 224)

Reviewer 3: This shape info might be confusing. Maybe it would be better to name it Numpy file shape

Author: This would require a change in grpc_serving_client.py script. Should I proceed?

Reviewer 3: Yes, that would be a minor change in the client

         0 Arctic fox, white fox, Alopex lagopus 279 ; Correct match.
Iteration 3; Processing time: 52.46 ms; speed 19.06 fps
imagenet top results in a single batch:
         0 bee 309 ; Correct match.
Iteration 4; Processing time: 51.36 ms; speed 19.47 fps
imagenet top results in a single batch:
         0 golden retriever 207 ; Correct match.
Iteration 5; Processing time: 46.52 ms; speed 21.50 fps
imagenet top results in a single batch:
         0 gorilla, Gorilla gorilla 366 ; Correct match.
Iteration 6; Processing time: 47.18 ms; speed 21.19 fps
imagenet top results in a single batch:
         0 magnetic compass 635 ; Correct match.
Iteration 7; Processing time: 48.54 ms; speed 20.60 fps
imagenet top results in a single batch:
         0 peacock 84 ; Correct match.
Iteration 8; Processing time: 48.72 ms; speed 20.53 fps
imagenet top results in a single batch:
         0 pelican 144 ; Correct match.
Iteration 9; Processing time: 48.77 ms; speed 20.51 fps
imagenet top results in a single batch:
         0 snail 113 ; Correct match.
Iteration 10; Processing time: 45.51 ms; speed 21.97 fps
imagenet top results in a single batch:
         0 zebra 340 ; Correct match.

> processing time for all iterations

Reviewer 3: use CLX host to show output results

Classification accuracy: 100.00
...
Start processing:
        Model name: resnet50DAG
        Iterations: 1
        Images numpy path: imgs.npy
        Images in shape: (10, 3, 224, 224)

Iteration 1; Processing time: 120.19 ms; speed 66.56 fps
imagenet top results in a single batch:
         0 airliner 404 ; Correct match.
         1 Arctic fox, white fox, Alopex lagopus 279 ; Correct match.
         2 bee 309 ; Correct match.
         3 golden retriever 207 ; Correct match.
         4 gorilla, Gorilla gorilla 366 ; Correct match.
         5 magnetic compass 635 ; Correct match.
         6 peacock 84 ; Correct match.
         7 pelican 144 ; Correct match.

processing time for all iterations
average time: 120.00 ms; average speed: 66.67 fps
median time: 120.00 ms; median speed: 66.67 fps
max time: 120.00 ms; min speed: 66.67 fps
min time: 120.00 ms; max speed: 66.67 fps
time percentile 90: 120.00 ms; speed percentile 90: 66.67 fps
time percentile 50: 120.00 ms; speed percentile 50: 66.67 fps
time standard deviation: 0.00
time variance: 0.00
Classification accuracy: 100.00
```
Each iteration presents the results of each infer request and details for each image in batch.

> With this feature we were able to successfully run two client scripts with different batch size parameters without reloading the model.

Reviewer 3: It should be mentioned that the clients can be started simultaneously without performance impact from the model reloading

> #### Checking metadata

Reviewer 3: Presenting the metadata might be useful before starting the client. It gives hints how the clients should sent the requests.
